### PR TITLE
Add mettle docker steps

### PR DESCRIPTION
**Pull request by adfoster-r7**

### CODE REVIEW

**Comment by smcintyre-r7** [comment]
The instructions didn't quite work for me. I included the output of my attempt below. What I noticed was that the first command didn't pull the docker image, which led to `rapid7/build` being pulled. Afterwards, the build failed.

<details>
<summary>Output</summary>

```
 : mettle: 15:30:55 mettle ~/Repositories/workshop/tools/docker-purge                                  
[*] Sleeping 10 seconds before purging all docker content, press Ctrl+C to abort...
[*] Stopping all containers... done.
[*] Removing all containers... done.
[*] Removing all images... done.
[*] Removing dangling volumes... done.
[*] Removing dangling networks... done.
  : mettle: 15:31:09 mettle docker pull rapid7/msf-ubuntu-x64-meterpreter:latest
latest: Pulling from rapid7/msf-ubuntu-x64-meterpreter
a4a2a29f9ba4: Already exists 
127c9761dcba: Already exists 
d13bf203e905: Already exists 
4039240d2e0b: Already exists 
04d4806b3c6d: Pull complete 
9ca04edc8278: Pull complete 
3f2f5473d8d5: Pull complete 
e53a57ff0449: Pull complete 
d1edee0d5490: Pull complete 
a9fa8cf74d05: Pull complete 
dba86be30972: Pull complete 
e130b5a02998: Pull complete 
8b1dc5210301: Pull complete 
6e83d1e12d66: Pull complete 
b19ce9dc083f: Pull complete 
2bcf1fb57f34: Pull complete 
Digest: sha256:a733f4ab29ed04a0e29e39e9e602efcbfc40512b3800f69078643ac13bd0ac77
Status: Downloaded newer image for rapid7/msf-ubuntu-x64-meterpreter:latest
docker.io/rapid7/msf-ubuntu-x64-meterpreter:latest
  : mettle: 15:35:38 mettle docker run -it --rm -v $(pwd):$(pwd) -w $(pwd) rapid7/build:mettle /bin/bash
Unable to find image 'rapid7/build:mettle' locally
mettle: Pulling from rapid7/build
6bbedd9b76a4: Pull complete 
fc19d60a83f1: Pull complete 
de413bb911fd: Pull complete 
2879a7ad3144: Pull complete 
668604fde02e: Pull complete 
e484dcac4a9f: Pull complete 
e1d382201857: Pull complete 
81a9da4660b4: Pull complete 
e3f78017ea92: Pull complete 
Digest: sha256:50884af3abd82fdfa32fee2b40da69934f5972793f9e05bb97a9bdb33feb6091
Status: Downloaded newer image for rapid7/build:mettle
root@a5f5dfe2bbe8:/home/smcintyre/mettle# make all-parallel -j 4 && rake mettle:check
Configuring automake
Building bison
make/Makefile.tools:175: recipe for target '/home/smcintyre/mettle/build/tools/bin/bison' failed
make: *** [/home/smcintyre/mettle/build/tools/bin/bison] Error 2
make: *** Waiting for unfinished jobs....
configure.ac:10: installing './compile'
configure.ac:3: installing './config.guess'
configure.ac:3: installing './config.sub'
configure.ac:4: installing './install-sh'
configure.ac:4: installing './missing'
extensions/sniffer/Makefile.am: installing './depcomp'
root@a5f5dfe2bbe8:/home/smcintyre/mettle# 
```
</details>

I've admitted found this process very difficult in the past and have just been relying on a script I wrote to automate the whole thing.

<details>
<summary>mettle-builder</summary>

```bash
#!/usr/bin/env bash
set -e

EX_USAGE=64
EX_NOINPUT=66

id="mettle-build"
directory="$(pwd)"
make_target="all-parallel"

params=()

show_help () {
  echo "usage: $(basename $0) [-h] [-d dir] [-t target]"
  echo ""
  echo "optional arguments:"
  echo "  -h, --help         show this help menu and exit"
  echo "  -d, --directory    build from the specified directory"
  echo "  -t, --target       build the specified make target"
  echo ""
  echo "example:"
  echo "  $(basename $0) -d mettle.pr"
  echo "  $(basename $0) -t TARGET=x86_64-linux-musl"
}

while (( "$#" )); do
  case "$1" in
    -h|--help)
      show_help
      exit 0
      ;;
    -d|--directory)
        if [ -n "$2" ] && [ ${2:0:1} != "-" ]; then
        if [ ! -d "$directory" ] ; then
          echo "Error: Argument for $1 must be a directory ($2)" >&2
          exit 1
        fi
        directory=$(realpath "$2")
        shift 2
      else
        echo "Error: Argument for $1 is missing" >&2
        exit 1
      fi
      ;;
    -t|--target)
      if [ -n "$2" ] && [ ${2:0:1} != "-" ]; then
        make_target=$2
        shift 2
      else
        echo "Error: Argument for $1 is missing" >&2
        exit 1
      fi
      ;;
    --)       # end argument parsing
      shift
      break
      ;;
    -*|--*=)  # unsupported flags
      printf "\e[1;31m[-]\e[0m Unsupported flag $1\n" >&2
      exit 1
      ;;
    *)        # preserve positional arguments
      params+=("$1")
      shift
      ;;
  esac
done

if [ "${#params[@]}" -ne 0 ]; then
  show_help
  exit 0
fi

if command -v getenforce &> /dev/null; then
  if getenforce | grep -i enforcing &> /dev/null; then
    printf "\e[1;33m[!]\e[0m The SELinux mode is currently: enforcing\n"
  fi
fi

if [ ! -f "$directory/Makefile" ]; then
  printf "\e[1;31m[-]\e[0m No Makefile was found in the target directory\n"
  exit $EX_NOINPUT
fi

if [[ "$make_target" =~ ^TARGET=(.*)$ ]]; then
  if ! grep -E "^${BASH_REMATCH[1]}$" "$directory/ARCHES" &> /dev/null; then
    printf "\e[1;31m[-]\e[0m Invalid target: ${BASH_REMATCH[1]}\n"
    exit $EX_USAGE
  fi
fi

printf "\e[1;34m[*]\e[0m Using directory: $directory\n"
printf "\e[1;34m[*]\e[0m Make target: $make_target\n"

if [[ "$(docker images -q $id 2> /dev/null)" == "" ]]; then
  printf "\e[1;34m[*]\e[0m Building the docker container...\n"
  docker build -t "$id" - <<-EOF
    FROM ubuntu:20.04
    ENV DEBIAN_FRONTEND noninteractive

    RUN apt-get update && \
        apt-get dist-upgrade -y && \
            apt-get -y install curl build-essential git autoconf automake libtool bison flex gcc ruby rake bundler git mingw-w64 && \
            apt-get clean && \
            rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
	EOF
fi

docker run \
  -it \
  --name "$id" \
  --mount type=bind,source="$directory",target=/mettle \
  --rm \
  -w /mettle \
  "$id" \
  /bin/bash -c "make $make_target -j 4 && rake build"
```
</details>

**Comment by adfoster-r7** [comment]
@smcintyre-r7 Thanks for running through this, I nuked all my local images, replicated the issue, and I've updated the readme to consistently use only `rapid7/msf-ubuntu-x64-meterpreter:latest` :+1:

**Comment by smcintyre-r7** [comment]
So now I'm getting build fails until after I install `curl` and `automake`. Those should have been included in the Docker image right?

File: README.md [hunk 1]
```diff
@@ -17,8 +17,32 @@ Debain, Ubuntu, and derivatives are most supported for builds. To build, you nee
...
 ```
 
-The Dockerfile under docker/Dockerfile contains a pre-configured build
-environment as well.
+Building on Docker
+------
+
+You can download a pre-built Docker image from [Rapid7's Docker Hub account](https://hub.docker.com/u/rapid7):
+```
+docker pull rapid7/msf-ubuntu-x64-meterpreter:latest
```
Comment (adfoster-r7) [inline README.md:25]: Looks like we're not using this, and we're relying on `rapid7/build:mettle`

That aside, the current Dockerfile isn't quite enough to run standalone without mounted volumes. Will close for now.

**Comment by adfoster-r7** [comment]
rapid7/build:mettle

**Pull request closed without merging**